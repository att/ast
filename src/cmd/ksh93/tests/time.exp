# Tests of `time` keyword and `times` alias.
set pid [spawn $ksh]
expect_prompt

set zero_sec_re "\[\t \]*0m0\.\[0-9\]\[0-9\]\[0-9\]s"
set one_sec_re "\[\t \]*0m1\.\[0-9\]\[0-9\]\[0-9\]s"

# ==========
log_test_entry
send "time\r"
expect -re "\r\nuser$zero_sec_re\r\nsys$zero_sec_re\r\n" {
    puts "time with no pipeline produces expected output"
}
expect_prompt

# ==========
# Timing a simple sleep command works.
log_test_entry
send "time sleep 1\r"
expect -re "\r\nreal\[\t \]*0m1\.00\[0-9\]s\r\nuser$zero_sec_re\r\nsys$zero_sec_re\r\n" {
    puts "time sleep 1 produces expected output"
}
expect_prompt

# ==========
# Timing a more complex pipeline works.
log_test_entry
send "time { sleep 0.05 | sleep 0.05; }\r"
expect -re "\r\nreal\[\t \]*0m0\.0\[0-9\]\[0-9\]s\r\nuser$zero_sec_re\r\nsys$zero_sec_re\r\n" {
    puts "time sleep 0.05 pipeline produces expected output"
}
expect_prompt

# ==========
# The times command with no args produces reasonable output.
log_test_entry
send "times\r"
expect -re "\r\n$zero_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected output"
}
expect_prompt

# ==========
# The times command with args produces a syntax error.
log_test_entry
send "times sleep 1\r"
expect "times: unexpected arguments\r\nUsage: times" {
    puts "times sleep 1 produces syntax error"
}
expect_prompt

# ==========
# Timing a CPU bound statement produces non-zero times.
log_test_entry
send "time { SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }\r"
expect -re "\r\nreal\[\t \]*$one_sec_re\r\nuser$one_sec_re\r\nsys$zero_sec_re\r\n" {
    puts "time CPU bound pipeline produces expected output"
}
expect_prompt

# ==========
# Start a CPU bound job in the background to force child user time to be non-zero.
log_test_entry
send "$ksh -c 'SECONDS=0; while (( SECONDS < 1.5 )); do true; done'\r"
expect_prompt

send "times\r"
# Note that we expect system mode times to be less than one but user mode time for the shell and
# child processes should be between one and two.
expect -re "\r\n$one_sec_re $zero_sec_re\r\n$one_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output"
}
expect_prompt
