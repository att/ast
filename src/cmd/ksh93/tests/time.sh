# Tests for time and times builtins

# Make sure all builtins are not enabled by default via PATH. Doing so makes `builtin -d`
# ineffective. Which breaks (at the time I write this) the final test in this module.
# See issue #960.
PATH=$NO_BUILTINS_PATH

# Make sure that we can just run these commands
time
times
