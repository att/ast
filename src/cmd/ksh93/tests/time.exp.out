time with no pipeline produces expected output
time sleep 1 produces expected output
time sleep 0.05 pipeline produces expected output
times produces expected output
times sleep 1 produces syntax error
time CPU bound pipeline produces expected output
times produces expected non-zero output
